### DISCUSSION TITLE
[HW3] I'm completely at a loss on how to approach this problem.

### DISCUSSION THREAD

**Discussion by marchmelo0923** [prosyslab-classroom/cs424-program-reasoning#431] [Q&A]
### Your Name

이신영

### Question

먼저, 이런 질문글을 올리게 된 것 정말 죄송합니다.

2번까지의 과제와는 다르게, 아예 접근법 자체를 찾지 못하고 있는 중이여서 질문하게 되었습니다.
다른 분들의 질의 글을 봤을 때는, 2번 과제와 유사한 방식으로 Z3 문법을 생성하고 이를 Z3.Fixedpoint.add_rule solver rule 를 통해, 각 R1, R2 ... 를 넣어야 하는 것으로 이해하고 있습니다.
그러나, llbasicblock에서 어떤식으로 각 명령어를 가져와야 하는지, 또한 relation이 정확히 무엇을 의미하여 어떤식으로 구성되어있고, 어떻게 접근해야 하는지에 대해 총체적으로 거의 이해하지 못하고 있는 상황입니다.

그렇기에, 불가피하게 이런 질문글을 올리게 되었습니다. 주어진 변수들을 활용할 방법에 대해, 시작점을 잡을 수 있도록 도와주셨으면 합니다.

아직까지도 감을 잘 못잡고 있는 상황이여서, 기한 내에 과제 제출이 어려울 수 있음에도, 이 과제에 접근하지 못하면 이후 과제도 난항이 있을 것 같아서 꼭 갈피를 잡고 싶습니다. 이런 식으로 적나라하게 과제에 대한 질문 올리게 된 점 다시 한 번 죄송합니다.

---
<Chatgpt translate>
First of all, I’m truly sorry for posting this question.

Unlike assignments 1 and 2, I am completely unable to figure out an approach for this task, which is why I decided to ask for help. From looking at other people’s questions, I understand that, similar to assignment 2, I should generate Z3 syntax and use Z3.Fixedpoint.add_rule with solver rule to add each R1, R2, and so on. However, I’m struggling to understand how to extract each instruction from llbasicblock, what exactly 'relation' means, how it’s structured, and how I should approach it.

For these reasons, I had no choice but to post this question. I would appreciate any guidance on how to use the given variables to get started.

I'm still having difficulty understanding, and though it may be challenging to submit this assignment on time, I feel it’s essential to get a grasp on this in order to manage future assignments. Once again, I apologize for posting such a direct question about the assignment.

**Comment by bongjunj** [comment]
Study the page 11 and 12 in the slides of the lecture 8.
Your job is to show the existence of a reachable state in the relation. Here, a relation corresponds to LLVM's basic block, parameterized by its free variables.

Few more hints:

* A rule is a first-order logic sentence. For example, `forall x. Entry() and x < 10 -> Cond(x)` is a rule you can add to the solver. Think about how to construct such rules from basic blocks you have.
* A relation is a set of states. Your verifier needs to show that the existence of reachable state in the target relation. For example, the `Entry` relation always has a reachable state so it is not parameterized by any variable. And, `Cond` relation is a set of reachable states  `x = 10`, `x = 9`, ..., so we can denote this as `Cond(10)`, `Cond(9)`, ....

**Reply by spearo2** [reply]
답변 감사합니다 @bongjunj 
위에서 말씀해주신 것처럼 과제를 접근하는 방법은 수업에서 다룬 내용과 동일합니다.
어떤 식으로 각 명령어를 가져와야 하는지에 대해서는 [llvmutils doc](https://prosyslab-classroom.github.io/llvmutils/)를 참고하시면 도움이 되실 것 같습니다.
포기하지 마시고 수업내용을 복습하다보면 답을 찾으실것이라 생각합니다!

============

Thank you for your response, @bongjunj.

As mentioned above, the approach to tackling the assignment aligns with the content covered in class. For guidance on how to utilize each llvm command, referring to the [llvmutils documentation](https://prosyslab-classroom.github.io/llvmutils/) may be helpful. Don't give up! I believe reviewing the class material will eventually lead you to the answer!

**Reply by marchmelo0923** [reply]
답변 주셔서 진심으로 감사합니다. 접근 방식 자체는 수업에서 배운것과 동일하다는 사실은 이해했고, forall 의 경우 Verifier의 make_rule에서 처리할 수 있음도 코드를 찾아보며 확인했습니다.

그러나, 여전히 llbasicblock을 다루는 방법에 대해서 잘 모르겠습니다.
참고로 주신 llvmutils doc 내의 함수에서도, llbasicblock을 이용하는 함수에 대한 내용을 찾지 못하고 있습니다. 괜찮으시다면 관련 내용에 대해서도 접근법을 여쭤봐도 되겠습니까?

감사하다는 말씀과 죄송하다는 말씀 같이 올립니다....

**Reply by sujin0529** [reply]
안녕하세요.
`llbasicblock` 타입 등을 다루는 함수들은 [llvm doc](https://llvm.moe/ocaml/Llvm.html) 에서 확인하실 수 있습니다. 과제 레포지토리 하단에 참고 자료로 llvm 을 다루는 방식 등이 제공되고 있으니 참고하시면 도움이 될 것 같습니다.
힘내세요!

감사합니다.

======================

Hello,
APIs dealing with `llbasicblock` type can be found at [llvm doc](https://llvm.mo/ocaml/Llvm.html). There is guidance to deal with llvm at the bottom of the homework repository. It would be helpful to refer to it.
Cheer up!

Thanks.

**Comment by livelykitten** [comment]
searching with class names such as llbasicblock and llmodule helps. C++ documentations have more detailed description, and it might look more familiar to you.

I looked at the code in hw1 to see how llbasicblock was handled, too.

llbasicblock이나 llmodule 등 클래스 이름으로 검색했을 때 도움이 되었습니다. C++ 문서인데, 더 자세한 설명이 있고, 익숙한 언어로 되어 있어 나름 도움이 된다 생각합니다.

저는 llbasicblock 다루는 법을 알아내던 도중 hw1의 코드를 참고하기도 했습니다.

https://llvm.org/doxygen/index.html

**Comment by marchmelo0923** [comment]
답변 감사합니다!!
많은 분들이 도와주신 덕분에 방향성을 잡은 것 같습니다
discussion은 닫아두겠습니다
